### PR TITLE
xinput: implemented XIChangeHierarchy

### PR DESCRIPTION
**Pull request by fkohlgrueber**
Hi,

First of all, thanks very much for this library and all the work you have invested in this project!

I'm getting more into xinput at the moment and I saw that the implementation of this extension isn't complete in python-xlib. I saw the chance to start contributing and so I implemented xlib's [XIChangeHierarchy](https://www.x.org/releases/X11R7.5/doc/man/man3/XIChangeHierarchy.3.html) function in xinput.py ;)

New functions:

``` python
dpy.xinput_add_master(name, send_core=True, enable=True)
dpy.xinput_remove_master(deviceid, return_mode=2, return_pointer=0, return_keyboard=0)
dpy.xinput_attach_slave(deviceid, new_master)
dpy.xinput_detach_slave(deviceid)
```

It's also possible to send multiple changes in a single request:

``` python
from Xlib.ext.xinput import ChangeHierarchyContext

with ChangeHierarchyContext(dpy) as c:
    c.add_master(...)
    c.remove_master(...)
    c.attach_slave(...)
    c.detach_slave(...)
```

You can use the `xinput` command line tool to see the device hierarchy and verify changes made by the above functions.

Unfortunately, I had to modify `rq.String8` because of an issue with the X Server expecting an zero-terminated string for `AddMaster`. The padding implementation didn't insert a termination character when entering a string that doesn't need padding (e.g. "test") which caused the X Server to crash. I made a minimal change (`pad=2`) to fix that. This change doesn't affect existing code, but I'm not sure if this problem can also occur with other functions. 

Ok, that's it. What do you think?

### CODE REVIEW

**Comment by vasily-v-ryabov** [comment]
Hi @caenyon, we appreciate contributions! Most of the code looks fine to me. I'm just not sure about `pad=2` hack. Shouldn't we count `slen` as `slen+1` with this new `\0` byte? Just to make padding correct.

**Comment by vasily-v-ryabov** [comment]
Good. I'm just thinking about separate parameter for String8 constructor: `String8('name', null_term=True)` that is more readable. @benoit-pierre ?

**Comment by benoit-pierre** [comment]
I honestly would prefer to wait for my changes for generating the extensions code, to avoid non-backward compatible API changes.

Regarding this particular implementation, I'm against changing the protocol for a bug in a version of the X11 server: the code works fine with `name="test"` on my machine without the `rq.String8('name', pad=2)` hack (Arch Linux, xorg-server 1.18.4-1).

**Comment by caenyon** [comment]
Ok, I just tried the sample again and it also worked on my machine when using "test". I also tried earlier versions of the code and found out that the crash was in fact caused by something else that I fixed in a later version. I'm sorry for the confusion. I will roll back the changes I did in rq.py in the next commit.

@benoit-pierre what do you mean by "generating the extensions code"? I'm just curious ;)

**Comment by benoit-pierre** [comment]
Generating the extension code: using the xcb-proto definitions and xcbgen.

File: Xlib/ext/xinput.py [hunk 1]
```diff
@@ -639,6 +639,123 @@ def ungrab_keycode(self, deviceid, keycode, modifiers):
...
+        self.events.append((RemoveMaster, deviceid, return_mode, return_pointer, return_keyboard))
+
+    def attach_slave(self, deviceid, new_master):
+        self.events.append((AttachSlave, deviceid, new_master))
+
+    def detach_slave(self, deviceid):
+        self.events.append((DetachSlave, deviceid))
+
+    def __exit__(self, *exc):
+        _change_hierarchy(self.display, *self.events)
```
Comment (benoit-pierre) [inline Xlib/ext/xinput.py]: This should not be called if a exception did occur:
```diff
diff --git i/Xlib/ext/xinput.py w/Xlib/ext/xinput.py
index 348f19f..fc2efdb 100644
--- i/Xlib/ext/xinput.py
+++ w/Xlib/ext/xinput.py
@@ -752,8 +752,9 @@ class ChangeHierarchyContext(object):
     def detach_slave(self, deviceid):
         self.events.append((DetachSlave, deviceid))

-    def __exit__(self, *exc):
-        _change_hierarchy(self.display, *self.events)
+    def __exit__(self, *exc_info):
+        if exc_info == (None, None, None):
+            _change_hierarchy(self.display, *self.events)


 def init(disp, info):
```

**Comment by drzraf** [comment]
@benoit-pierre could you please provide the method/template/script you used to generate the bindings from xcb-proto XML files as you did in #42?

**Comment by benoit-pierre** [comment]
@drzraf: script here: https://gist.github.com/de5beb2cfb1a60abe78f83619796edea.

I have not updated it in some time, so it won't work with the latest version of xcb-proto, you'll need to use version 1.11:
```shell
$ git clone git://anongit.freedesktop.org/git/xcb/proto
$ git -C proto checkout 1.11
$ PYTHONPATH="$PWD/proto" python3 ./gen_xlib.py proto/src/xinput.xml
```
This will output the generated code on stdout. Note that this might generate some code that is not compatible with the current python-xlib master: I have an (old) branch with a lot of changes to support things like unions, switches, variable length fields with complex length calculation, etc...

**Pull request closed without merging**